src/Raven.Client/Documents/Session/TimeSeries/TimeSeriesEntry.cs

```diff
@@ -28,8 +33,89 @@ public abstract class TimeSeriesAggregatedEntry : TimeSeriesEntryValues
 
     }
 
-    public abstract class TimeSeriesEntryValues
+    [AttributeUsage(AttributeTargets.Field)]
+    public class TimeSeriesValue : Attribute
+    {
+        public int Index;
+        public TimeSeriesValue(int index)
+        {
+            Index = index;
+        }
+    }
+
+    public abstract class TimeSeriesEntryValues : IPostDeserialization
     {
         public double[] Values { get; set; }
+
+        private static readonly ConcurrentDictionary<Type, TimeSeriesEntryMappingInfo> _cache = new ConcurrentDictionary<Type, TimeSeriesEntryMappingInfo>();
+
+        private class TimeSeriesEntryMappingInfo
+        {
+            public readonly SortedDictionary<int, FieldInfo> FieldsMapping = new SortedDictionary<int, FieldInfo>();
+        }
+
+        internal void SetValuesFromFields()
+        {
+            var t = GetType();
+            var mapping = GetFieldsMapping(t);
+            if (mapping == null)
+                return;
+
+            var fields = mapping.FieldsMapping;
+            Values = new double[fields.Count];
+            foreach (var fieldInfo in fields)
+            {
+                var index = fieldInfo.Key;
+                var field = fieldInfo.Value;
+                Values[index] = (double)field.GetValue(this);
+            }
+        }
+
+        private static TimeSeriesEntryMappingInfo GetFieldsMapping(Type t)
+        {
+            if (_cache.TryGetValue(t, out var mapping) == false)
+            {
+                foreach (var field in t.GetFields())
+                {
+                    var attribute = (TimeSeriesValue)field.GetCustomAttributes(false).SingleOrDefault(a => a is TimeSeriesValue);
+                    if (attribute == null)
+                        continue;
+
+                    var i = attribute.Index;
+                    mapping ??= new TimeSeriesEntryMappingInfo();
+                    mapping.FieldsMapping[i] = field;
+                }
+
+                mapping = _cache.GetOrAdd(t, mapping);
```

Comment:
`mapping = _cache.GetOrAdd(t, CreateMapping);` would be enough. Do not need to use TryGet and then GetOrAdd

Reply:
Not sure I'm following.. This should be thread-safe...